feat(worker): seal block early if we're over target block time

Thegaram opened the pull request:
## 1. Purpose or design rationale of this PR

If we're over the target block time, short-circuit block packing and leave transactions to the next block.

This is a short-term fix to make the fork block more predictable.

## 2. PR title

Your PR title must follow [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) (as we are doing squash merge for each PR), so it must start with one of the following [types](https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type):

- [X] feat: A new feature


## 3. Deployment tag versioning

Has the version in `params/version.go` been updated?

- [ ] This PR doesn't involve a new deployment, git tag, docker image tag, and it doesn't affect traces
- [X] Yes


## 4. Breaking change label

Does this PR have the `breaking-change` label?

- [X] This PR is not a breaking change
- [ ] Yes

Thegaram committed:
feat(worker): seal block early if we're over target block time

omerfirmak commented on miner/worker.go:
```diff
@@ -1055,6 +1059,10 @@ loop:
 			return atomic.LoadInt32(interrupt) == commitInterruptNewHead, circuitCapacityReached
 		}
+		// seal block early early if we're over time
+		if w.current.tcount > 0 && w.chainConfig.Clique != nil && uint64(time.Now().Unix()) > w.current.parentTimestamp+w.chainConfig.Clique.Period {
```
I don't think we need to use `parentTimestamp` here, `w.current.header.Timestamp` is already derived from `parentTimestamp`.

We can simply check `time.Now() > current.header.Timestamp`

Thegaram replied:
You mean because we set `header.Time` it here?

https://github.com/scroll-tech/go-ethereum/blob/d81a3308b18232c134335a6cfe59cf3df0611658/consensus/clique/clique.go#L557-L560

I guess I was worried about the case where `parentTimestamp + clique.Period < time.Now()` and so `current.header.Timestamp == time.Now()`. But it's probably not an issue, if line 559 is triggered we're already over time and whenever `time.Now() > current.header.Timestamp` we can seal the block.

I'll update.

Thegaram replied:
Updated in [29a3cb9](https://github.com/scroll-tech/go-ethereum/pull/706/commits/29a3cb925481e07d75832641bad80aa7e1b933a4).

0xmountaintop replied:
LGTM will approve once it's updated

0xmountaintop commented on miner/worker.go:
```diff
@@ -1055,6 +1059,10 @@ loop:
 		}
+		// seal block early early if we're over time
+		if w.current.tcount > 0 && w.chainConfig.Clique != nil && uint64(time.Now().Unix()) > w.current.parentTimestamp+w.chainConfig.Clique.Period {
+			return false, true
```
why not `return false, circuitCapacityReached`? then we can reflect `circuitCapacityReached` more accurately?

omerfirmak replied:
returning a constant `true`forces the caller to seal the block.

Thegaram replied:
Yes, that was my thinking. Skip subsequent invocations of `commitTransactions` inside `commitNewWork`.

georgehao commented on miner/worker.go:
```diff
@@ -1055,6 +1059,10 @@ loop:
 			}
 			return atomic.LoadInt32(interrupt) == commitInterruptNewHead, circuitCapacityReached
 		}
+		// seal block early early if we're over time
```
```suggestion
		// seal block early if we're over time
```

Thegaram committed:
use current.header.Time

Thegaram committed:
typo

omerfirmak commented:
Small q; do empty blocks need any special treatment?

0xmountaintop commented:
we return instead of break in the loop. are we comfortable with skipping

```
	if !w.isRunning() && len(coalescedLogs) > 0 {
		// We don't push the pendingLogsEvent while we are mining. The reason is that
		// when we are mining, the worker will regenerate a mining block every 3 seconds.
		// In order to avoid pushing the repeated pendingLog, we disable the pending log pushing.

		// make a copy, the state caches the logs and these logs get "upgraded" from pending to mined
		// logs by filling in the block hash when the block was mined by the local miner. This can
		// cause a race condition if a log was "upgraded" before the PendingLogsEvent is processed.
		cpy := make([]*types.Log, len(coalescedLogs))
		for i, l := range coalescedLogs {
			cpy[i] = new(types.Log)
			*cpy[i] = *l
		}
		w.pendingLogsFeed.Send(cpy)
	}
	// Notify resubmit loop to decrease resubmitting interval if current interval is larger
	// than the user-specified one.
	if interrupt != nil {
		w.resubmitAdjustCh <- &intervalAdjust{inc: false}
	}
	return false, circuitCapacityReached
```

Thegaram committed:
break instead of return

Thegaram commented:
> we return instead of break in the loop. are we comfortable with skipping

Good question. This code does not seem super relevant, but skipping it might return in some difference in how we expose pending blocks. Changed to `break` in [91627fb](https://github.com/scroll-tech/go-ethereum/pull/706/commits/91627fb51fbb2c31fdf20d90591e7b6aa9874add) to be more consistent with the original behaviour.

> Small q; do empty blocks need any special treatment?

I added the `w.current.tcount > 0` condition to make sure we always make some progress processing transactions. It is possible that we keep processing and skipping transactions, but this does not seem to happen often in practice.
